src/GraphQL/Subscription/ObservableExtensions.cs

```diff
@@ -0,0 +1,253 @@
+
+            public void OnCompleted() => QueueEvent(QueueType.Completion, null, null);
+
+            /// <summary>
+            /// Queues the specified event and if necessary starts watching for an event to complete.
+            /// </summary>
+            private void QueueEvent(QueueType queueType, Task<TOut>? task, Task<Exception>? error)
+            {
+                var queueData = new QueueData { Type = queueType, Data = task, Error = error };
+                bool attach = false;
+                lock (_queue)
+                {
+                    _queue.Enqueue(queueData);
+                    attach = _queue.Count == 1;
+                }
+
+                // start watching for an event to complete, if this is the first in the queue
+                if (attach)
+                {
+                    // for data events, start sending the data once the transformation task completes
+                    // but for error/completion events, send the event now
+                    if (task != null)
+                    {
+                        // start returning data once the first task has completed (or now if already completed)
+                        if (task.IsCompleted)
+                        {
+                            _ = ReturnDataAsync();
+                        }
+                        else
+                        {
+                            _ = task.ContinueWith(_ => ReturnDataAsync());
+                        }
+                    }
+                    else
+                        // start returning data now
+                        _ = ReturnDataAsync();
+                }
+            }
+
+            /// <summary>
+            /// Returns data from the queue in order (or raises errors or completed notifications);
+            /// executes until the queue is empty.
+            /// </summary>
+            private async Task ReturnDataAsync()
+            {
+                // grab the event at the start of the queue, but don't remove it from the queue
+                QueueData? queueData;
+                lock (_queue)
+                {
+                    // should always successfully peek from the queue here
+                    queueData = _queue.Count > 0 ? _queue.Peek() : null;
+                }
+                while (queueData != null)
+                {
+                    // process the event
+                    if (queueData.Type == QueueType.Data)
+                    {
+                        await ProcessDataAsync(queueData.Data!).ConfigureAwait(false);
+                    }
+                    else if (queueData.Type == QueueType.Error)
+                    {
+                        await ProcessErrorAsync(queueData.Error!).ConfigureAwait(false);
+                    }
+                    else if (queueData.Type == QueueType.Completion)
+                    {
+                        if (!_token.IsCancellationRequested)
+                            _observer.OnCompleted();
```

Comment:
🤔 May be move these two lines of code into separate method like `ProcessCompletion` to align with other two. Yeah, just cosmetic change and no async/configureawait, as you see fit.

Reply:
Done